network_simulator/host.py

```diff
@@ -23,4 +28,91 @@ def __init__(self, address=None, flows={}, flow_packets_received={},
+    """
+
+    def __init__(self, host, packet, packet_previously_received):
+        """
+        Sanity check: make sure destination was this Host and that this isn't
+        a RouterPacket.
+        :ivar host: Host that receives Packet
+        :ivar packet: Packet received
+        :ivar packet_previously_received: True if Packet previously received
+        """
+        Event.__init__(self)
+
+        # Sanity check
+        assert packet.dest_id == host.address and not isinstance(packet,
+                                                                 RouterPacket)
+        self.host = host
+        self.packet = packet
+        self.packet_previously_received = packet_previously_received
+
+    def elem_in_list(elem, my_list):
+        # ind is where this element *would* be inserted
+        # into the list in order to keep it in sorted order.
+        ind = bisect.bisect_left(my_list, elem)
+        if ind != len(my_list) and my_list[ind] == elem:
+            return True
+        else:
+            return False
+
+    def run(self, main_event_loop, statistics):
+        """
+        Put packet in flow_packets_received and sort.
+        Update statistics.
+        :param main_event_loop: event loop where new Events will be scheduled.
+        :param statistics: statistics to update
+        """
+        # add packet to host.flow_packets_received
+        old_packets_received = self.host.flow_packets_received.get(
+            self.packet.flow_id, default=list())
+
+        self.packet_previously_received = elem_in_list(packet.packet_id,
+                                                       old_packets_received)
+        if not self.packet_previously_received:
+            bisect.insort(old_packets_received, self.packet.packet_id)
+            self.host.flow_packets_received[self.packet.flow_id] = \
+                old_packets_received
+
+        else:
+            pass
+
+        statistics.host_packet_received(self.host, self.packet,
+                                        main_event_loop.global_clock_sec)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule FlowReceivedAckEvent.
+        Populate AckPacket for selective repeat.
+
+        If packet was data packet, generate Ack and schedule via
+        DeviceToLink event unless packet already received earlier.
+
+        AckPackets, set original data packet's start time for RTT.
+        AckPacket ID same as DataPacket ID
+        :param main_event_loop: schedule new Events
+        """
+
+
+        if isinstance(self.packet, AckPacket):
+            # look up the flow associated with this AckPacket.
+            f = FlowReceivedAckEvent(self.flows[self.packet.flow_id],
+                                     self.packet)
+            main_event_loop.schedule_event_with_delay(f, 0.0)
+
+        if isinstance(self.packet, DataPacket) and \
+                        self.packet_id not in self.flow_packets_received:
```

Comment:
the second line of this conditional should be "and not self.packet_previously_received". Remember that flow_packets_received is a map from flow ID to lists, so your current check does not work (it amounts to checking if packet_id is a key in flow_packets_received, which it cannot be because the keys to this map are *flow IDs*).